whats new - seeding (#4836)
dotnet/EntityFramework.Docs: entity-framework/core/modeling/data-seeding.md

ms.date: 11/02/2018
uid: core/modeling/data-seeding
---

> 

Reviewer: Update date

# Data Seeding

Data seeding is the process of populating a database with an initial set of data.

There are several ways this can be accomplished in EF Core:

* Configuration options data seeding (`UseSeeding`)
* Custom initialization logic
* Model managed data (`HasData`)

> * Manual migration customization

Reviewer: Add links pointing to the anchors below

<a name="use-seeding-method"></a>

## Configuration options `UseSeeding` and `UseAsyncSeeding` methods

> EF 9 introduced `UseSeeding` and `UseAsyncSeeding` methods, which provide a convenient way of seeding the database with initial data. These methods aim to improve the experience of using custom initialization logic (explained below). They provide one clear location where all the data seeding code can be placed. Moreover, the code inside `UseSeeding` and `UseAsyncSeeding` methods is protected by the [migration locking mechanism](/ef/core/what-is-new/ef-core-9.0/whatsnew#concurrent-migrations) to prevent concurrency issues. The new seeding methods are called as part of `EnsureCreated` operation, `DbContext.Database.Migrate()` and `dotnet ef database update` command, even if there are no model changes and no migrations were applied.

Reviewer: ```suggestion
EF 9 introduced `UseSeeding` and `UseAsyncSeeding` methods, which provide a convenient way of seeding the database with initial data. These methods aim to improve the experience of using custom initialization logic (explained below). They provide one clear location where all the data seeding code can be placed. Moreover, the code inside `UseSeeding` and `UseAsyncSeeding` methods is protected by the [migration locking mechanism](/ef/core/what-is-new/ef-core-9.0/whatsnew#concurrent-migrations) to prevent concurrency issues and are now the recommended way to add initial data in your app.

The new seeding methods are called as part of `<xref:Microsoft.EntityFrameworkCore.Storage.IDatabaseCreator.EnsureCreated>` operation, `<xref:Microsoft.EntityFrameworkCore.RelationalDatabaseFacadeExtensions.Migrate>` and `dotnet ef database update` command, even if there are no model changes and no migrations were applied.
```

There are several ways this can be accomplished in EF Core:

* Model managed data
* Manual migration customization
* Custom initialization logic

> * Configuration options seeding methods

Reviewer: Agree, and concerning the renaming of the feature this is the place where we should explain that it was referred to as "seeding" before, but this name set up incorrect expectations, that's why the new name does not contain "seeding".

As to the order of preference. It should be:

1. UseSeeding
2. Custom initialization logic (only if the above is not enough, e.g. you need to seed using two contexts in one transaction)
3. HasData
4. Manual migration customization (if the above is not enough)

> * Configuration options seeding methods

Reviewer: I still don't see a sentence saying that the feature name was changed

> * Configuration options seeding methods

Author: > [!NOTE]
> Populating the database using the `HasData` method used to be referred to as "data seeding". This naming sets incorrect expectations, as the feature has a number of limitations and is only appropriate for specific types of data. That is why we decided to rename it to "model managed data". `UseSeeding` and `UseAsyncSeeding` methods should be used for general purpose data seeding.

> * Configuration options seeding methods

Author: added it in the section for HasData rather than on top

## Model managed data

In EF Core, data can be associated with an entity type as part of the model configuration. Then, EF Core [migrations](xref:core/managing-schemas/migrations/index) can automatically compute what insert, update or delete operations need to be applied when upgrading the database to a new version of the model.

> [!NOTE]
Migrations only considers model changes when determining what operation should be performed to get the managed data into the desired state. Thus any changes to the data performed outside of migrations might be lost or cause an error.

As an example, this will configure managed data for a `Country` in `OnModelCreating`:

[!code-csharp[CountrySeed](../../../samples/core/Modeling/DataSeeding/ManagingDataContext.cs?name=CountrySeed)]

To add entities that have a relationship the foreign key values need to be specified:

[!code-csharp[CitySeed](../../../samples/core/Modeling/DataSeeding/ManagingDataContext.cs?name=CitySeed)]

> When managing data for Many-to-Many navigations, the join entity needs to be defined explicitly. If the entity type has any properties in shadow state (e.g. the `LanguageCountry` join entity below), an anonymous class can be used to provide the values:

Reviewer: ```suggestion
When managing data for many-to-many navigations, the join entity needs to be configured explicitly. If the entity type has any properties in shadow state (e.g. the `LanguageCountry` join entity below), an anonymous class can be used to provide the values:
```

Alternatively, you can use `context.Database.EnsureCreated()` to create a new database containing the managed data, for example for a test database or when using the in-memory provider or any non-relational database. Note that if the database already exists, `EnsureCreated()` will neither update the schema nor managed data in the database. For relational databases you shouldn't call `EnsureCreated()` if you plan to use Migrations.

### Limitations of model managed data

This type of data is managed by migrations and the script to update the data that's already in the database needs to be generated without connecting to the database. This imposes some restrictions:

* The primary key value needs to be specified even if it's usually generated by the database. It will be used to detect data changes between migrations.
* Previously managed data will be removed if the primary key is changed in any way.

Therefore this feature is most useful for static data that's not expected to change outside of migrations and does not depend on anything else in the database, for example ZIP codes.

> If your scenario includes any of the following it is recommended to use custom initialization logic (or `UseSeeding` and `UseAsyncSeeding` methods) described in the last section:

Reviewer: ```suggestion
If your scenario includes any of the following it is recommended to use custom initialization logic (or `UseSeeding` and `UseAsyncSeeding` methods) described in the first section:
```

> [!NOTE]
> Populating the database using the `HasData` method used to be referred to as "data seeding". This naming sets incorrect expectations, as the feature has a number of limitations and is only appropriate for specific types of data. That is why we decided to rename it to "model managed data". `UseSeeding` and `UseAsyncSeeding` methods should be used for general purpose data seeding.

### Limitations of model managed data

This type of data is managed by migrations and the script to update the data that's already in the database needs to be generated without connecting to the database. This imposes some restrictions:

* The primary key value needs to be specified even if it's usually generated by the database. It will be used to detect data changes between migrations.

> * Previously managed data will be removed if the primary key is changed in any way.

Reviewer: ```suggestion
* Previously inserted managed data will be removed if the primary key is changed in any way.
```

* Deploying the initialization app with the main app, invoking the initialization routine and disabling or removing the initialization app.

This can usually be automated by using [publish profiles](/aspnet/core/host-and-deploy/visual-studio-publish-profiles).

<a name="use-seeding-method"></a>

### Configuration options `UseSeeding` and `UseAsyncSeeding` methods

> EF 9 introduced `UseSeeding` and `UseAsyncSeeding` methods, which provide convenient way of configuring custom initialization logic. These methods can be set up in the [options configuration step](/ef/core/dbcontext-configuration/#dbcontextoptions). Here is an example analogous to the custom initialization showed above:

Reviewer: Also mention that one of these methods is called when `Migrate` is called or when `dotnet ef database update` is run, even if no migrations were applied

[!code-csharp[LanguageSeed](../../../samples/core/Modeling/DataSeeding/DataSeedingContext.cs?name=ContextOptionSeeding)]

> !NOTE

> > `UseSeeding` is called from the `EnsureCreated` method, and `UseAsyncSeeding` is called from the `EnsureCreatedAsync` method.  When using this feature, it is recommended to implement both `UseSeeding` and `UseAsyncSeeding` methods using similar logic. This way the seeding code runs for synchronous and asynchrouns scenarios.

Reviewer: This is a safeguard, as the tooling will only call the sync version. At least until https://github.com/dotnet/efcore/issues/34923 is implemented

> > `UseSeeding` is called from the `EnsureCreated` method, and `UseAsyncSeeding` is called from the `EnsureCreatedAsync` method.  When using this feature, it is recommended to implement both `UseSeeding` and `UseAsyncSeeding` methods using similar logic. This way the seeding code runs for synchronous and asynchrouns scenarios.

Reviewer 2: Ah OK, I forgot about the tooling problem. Should we maybe be more explicit, saying that (currently) sync `UseSeeding` **must** be implemented in order for the tooling to work? The current wording makes it a bit unclear why/when...
